File: legacy/src/main/rs/emulate/legacy/map/MapIndexDecoder.java
package rs.emulate.legacy.map;

import rs.emulate.legacy.archive.Archive;
import rs.emulate.shared.util.DataBuffer;

import java.io.FileNotFoundException;

/**
 * Decoder for the version lists map-index entry.
 */
public class MapIndexDecoder {
    private final DataBuffer data;

    /**
     * Construct a new MapIndexDecoder.
     *
     * @param versionListArchive The version list archive
     * @throws FileNotFoundException
     */
    public MapIndexDecoder(Archive versionListArchive) throws FileNotFoundException {
        this.data = versionListArchive.getEntry("map_index").getBuffer().asReadOnlyBuffer();
    }

    public MapIndex decode() {
        int areaCount = data.remaining() / 7;

        int[] areas = new int[areaCount];
        int[] landscapes = new int[areaCount];
        int[] mapFiles = new int[areaCount];
        int[] membersArea = new int[areaCount];

        for (int area = 0; area < areaCount; area++) {
            areas[area] = data.getUnsignedShort();
            mapFiles[area] = data.getUnsignedShort();
            landscapes[area] = data.getUnsignedShort();
            membersArea[area] = data.getUnsignedByte();
        }

        return new MapIndex(areas, landscapes, mapFiles, membersArea);
    }
}

File: legacy/src/main/kotlin/rs/emulate/legacy/version/map/MapIndexDecoder.kt
package rs.emulate.legacy.version.map

import rs.emulate.legacy.archive.Archive
import java.io.FileNotFoundException

/**
 * Decoder for the version lists "map_index" entry.
 *
 * @param archive The version list [Archive].
 * @throws FileNotFoundException If the `map_index` entry could not be found.
 */
class MapIndexDecoder constructor(archive: Archive) {

    private val data = archive[MapIndex.ENTRY_NAME].buffer.copy()

    /**
     * Decodes the contents of the `map_index` entry into a [MapIndex].
     */
    fun decode(): MapIndex {
        val count = data.readableBytes() / (3 * java.lang.Short.BYTES + java.lang.Byte.BYTES)

        val areas = IntArray(count)
        val maps = IntArray(count)
        val objects = IntArray(count)
        val members = BooleanArray(count)

        for (index in 0 until count) {
            areas[index] = data.readUnsignedShort()
            maps[index] = data.readUnsignedShort()
            objects[index] = data.readUnsignedShort()
            members[index] = data.readBoolean()
        }

        return MapIndex(areas, objects, maps, members)
    }

}
